lib/compress/zstd_compress.c

```diff
@@ -4461,6 +4486,400 @@ size_t ZSTD_compress2(ZSTD_CCtx* cctx,
+            } else {
+                /* This sequence ends inside the literals, break to store the last literals */
+                break;
+            }
+        }
+        /* Check if this offset can be represented with a repcode */
+        {   U32 ll0 = (litLength == 0);
+            offCode = ZSTD_finalizeOffCode(rawOffset, updatedRepcodes.rep, ll0);
+            updatedRepcodes = ZSTD_updateRep(updatedRepcodes.rep, offCode, ll0);
+        }
+
+        seqPos->posInSrc += litLength + matchLength;
+        FORWARD_IF_ERROR(ZSTD_validateSequence(offCode, matchLength, seqPos->posInSrc,
+                                               cctx->appliedParams.cParams.windowLog, dictSize),
+                         "Sequence validation failed");
+        DEBUGLOG(6, "Storing sequence: (of: %u, ml: %u, ll: %u)", offCode, matchLength, litLength);
+        ZSTD_storeSeq(seqStore, litLength, ip, iend, offCode, matchLength - MINMATCH);
+        ip += matchLength + litLength;
+    }
+    DEBUGLOG(5, "Ending seq: idx: %u (of: %u ml: %u ll: %u)", idx, inSeqs[idx].offset, inSeqs[idx].matchLength, inSeqs[idx].litLength);
+    assert(idx == inSeqsSize || endPosInSequence <= inSeqs[idx].litLength + inSeqs[idx].matchLength);
+    seqPos->idx = idx;
+    seqPos->posInSequence = endPosInSequence;
+    ZSTD_memcpy(cctx->blockState.nextCBlock->rep, updatedRepcodes.rep, sizeof(repcodes_t));
+
+    iend -= bytesAdjustment;
+    if (ip != iend) {
+        /* Store any last literals */
+        U32 lastLLSize = (U32)(iend - ip);
+        assert(ip <= iend);
+        DEBUGLOG(6, "Storing last literals of size: %u", lastLLSize);
+        ZSTD_storeLastLiterals(seqStore, ip, lastLLSize);
+        seqPos->posInSrc += lastLLSize;
+    }
+
+    return bytesAdjustment;
+}
+
+/* Compress, block-by-block, all of the sequences given.
+ *
+ * Returns the cumulative size of all compressed blocks (including their headers), otherwise a ZSTD error.
+ */
+static size_t ZSTD_compressSequences_internal(void* dst, size_t dstCapacity,
+                                              ZSTD_CCtx* cctx,
+                                              const ZSTD_Sequence* inSeqs, size_t inSeqsSize,
+                                              const void* src, size_t srcSize) {
+    size_t cSize = 0;
+    U32 lastBlock;
+    size_t blockSize;
+    size_t compressedSeqsSize;
+    size_t remaining = srcSize;
+    ZSTD_sequencePosition seqPos = {0, 0, 0};
+    U32 repCodesBackup[ZSTD_REP_NUM];   /* If we emit block as nocompress or RLE block, the decoder will
+                                         * never "see" those sequences. So, in order for the repcode table
+                                         * to remain in sync, we need to revert the repcode table back to the
+                                         * state it was in before processing the sequences in the RLE or nocompress
+                                         * block.
+                                         */
```

Comment:
I believe this should fix the edge case where RLE blocks (and I'm guessing uncompressed blocks) ruin repcode history for the next three matches.

If I understand correctly, the reason that RLE blocks (and no compress blocks) are troubling to the current API is that on the decoder side, it never sees the sequences that might have been inside an RLE block or nocompress block. But the issue is that in this API we update the repcode table using those "unused" sequences anyways.

To mitigate, we keep a copy of the repcodes that this current block would start with. After `ZSTD_copySequencesToSeqStore`, the block will end up with a new repcode table in the `blockState`. If it's the case that we emit the block as RLE or noCompress, then we copy the repcodes from this backup back into the `blockState`. Does this make sense?

Reply:
Actually, looking at my code again, I actually think this case might have already been taken care of. Since `ZSTD_confirmRepcodesAndEntropyTables()` is only called when we emit a compressed block. That means when we emit an RLE or noCompress block, we don't actually update the value of `rep` in `prevCBlock`. 

Therefore, when we begin processing the sequences for the next block, it's actually using the same starting repcode table values as the block that was emitted uncompressed, or RLE.